source/AI.cpp

```diff
@@ -4582,7 +4606,10 @@ void AI::IssueOrders(const PlayerInfo &player, const Orders &newOrders, const st
 			// asteroid.
 			if(hasMismatch && targetAsteroid)
 				alreadyHarvesting = (existing.type == newOrders.type) && (newOrders.type == Orders::HARVEST);
-			existing = newOrders;
+
+			// The hold fire order should not overwrite compatible old orders like movements.
+			if(newOrders.type != Orders::HOLD_FIRE || (existing.type == Orders::FINISH_OFF || existing.type == Orders::ATTACK))
+				existing = newOrders;
```

Comment:
This still makes the holdFire set necessary. I was thinking something more like this after line 4600. (Not guaranteeing that this will work exactly as written.)
```c++
// If the existing orders are to hold fire, continue to hold fire.
if(existing.type & Orders::HOLD_FIRE)
	newOrders.type |= Orders::HOLD_FIRE;
// If the new order is to hold fire, then maintain all old orders.
if(newOrders.type == Orders::HOLD_FIRE)
	newOrders.type |= existing.type;
// Holding fire is mutually exclusive with attacking.
if((newOrders.type & Orders::HOLD_FIRE) && ((newOrders.type & Orders::FINISH_OFF) || (newOrders.type & Orders::ATTACK)))
{
	// If we were holding fire before, stop holding fire.
	if(existing.type & Orders::HOLD_FIRE)
		newOrders.type -= Orders::HOLD_FIRE;
	// If we were attacking before, stop attacking.
	else
	{
		if(existing.type & Orders::FINISH_OFF)
			newOrders.type -= Orders::FINISH_OFF;
		if(existing.type & Orders::ATTACK)
			newOrders.type -= Orders::ATTACK;
	}
}
```

Reply:
orders can only contain one order at a time, right? because then you cant combine hold fire with smth like hold position, can you?

Reply:
> orders can only contain one order at a time, right? because then you cant combine hold fire with smth like hold position, can you?

Orders is really just an integer where we're flipping specific bits depending on the command that we give to the escorts. There's not actually anything stopping an Order from containing multiple orders at once; we just need to make sure that we're checking if a specific bit is set instead of checking if the order.type equals a specific order constant.

> (also if we get more orders like this it'd get even more messy, wouldnt it?)

It would get messy with the method that I commented with (which was admittedly just me throwing something together that I thought would work rather than actually spending the time to make something that's good), but we can streamline things if we take advantage of bitwise operations.

Imagine that order.type is just four bits: 0000. Reading left to right, let's say the bits represent hold fire, focus fire, hold position, and gather. If we want to toggle whether the escorts are gathering around the flagship then we toggle back and forth between XXX1 and XXX0. But focus fire and hold position are mutually exclusive with gathering, so we're actually flipping between X001 and XXX0. Instead of having a bunch of if statements that check different order constant, each order could instead have a bitmask that determines which orders it is mutually exclusive with. So in the case of gather, the gather constant would be 0001 and its bitmask would be 1001. To apply the order, we'd then do `(order.type ^ 0001) & 1001` whenever the gather command is sent. `order.type ^ 0001`, where ^ is a xor, lets us toggle the gather order bit, then doing `& 1001` means that the hold fire bit remains what it is, the focus fire and hold position bits get set to 0, and the gather bit remains what it is.

Expanding this to the other possible commands, we'd get the following:
Command | Constant | Bitmask
---|---|---
Gather | 0001 | 1001 (clears hold position and focus fire but leaves hold fire)
Hold Position | 0010 | 1010 (clears focus fire and gather but leaves hold fire)
Focus Fire | 0100 | 0100 (clears everything but itself)
Hold Fire | 1000 | 1011 (clears focus fire)

We already have a Bitset class used by Personality that could potentially streamline some of this instead of keeping order.type as a bare int. Might need some new functions for the bitmasking, though, as I don't see any way to easily apply a bitmask to the bitset.

Reply:
yeah thats what I was thinking as a "proper" way (although idk how much it scales with more orders - more bits ig), the thing is it'd probably take a bit of work and could lead to bugs if not done right ~~I was fine with a holdFire set~~

Reply:
> the thing is it'd probably take a bit of work and could lead to bugs if not done right

Well that's true of a lot of things. Doesn't mean we don't do them. 

The holdFire set works for now, but what if we add another order that isn't mutually exclusive with other orders? Using bitmasks is explicitly designed to be able to handle that case, while not doing so probably means having another set for each additional order.

Reply:
Since this would be a refactoring of Orders that could be done on its own, I could give it a shot. 🤔

Reply:
Actually, we could just use a std::bitset instead of our implementation.

Reply:
Yeah and then no need to worry about what is compatible with what for every single command, right? Mb we could add a function to remove those incompatibilities and make the code more smooth at the same time
I dont quite have the time for that in the near future unfortunately so if you do, please go ahead

Reply:
Heads up that I have this refactor in the works. Mainly just need to test it. Currently focusing on other PRs, though.

Reply:
ok great I wont touch anything else then, testing is precisely the thing I dont quite have time to do haha